src/coreclr/jit/morph.cpp

```diff
@@ -11161,6 +11161,11 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
             //
             if (!tree->AsIndir()->IsVolatile())
             {
+                if (op1->IsIconHandle(GTF_ICON_STR_HDL))
+                {
+                    tree->gtFlags |= (GTF_IND_INVARIANT | GTF_IND_NONFAULTING | GTF_IND_NONNULL);
+                }
```

Comment:
What is introducing this node without setting the flags? Wouldn't it be better to fix it at that point so it's the same as all other places that create these indirs?

Reply:
Ah sorry, I just reread the PR description. Still, it would be nice to know exactly how this node appears -- if this is due to general folding/propagation, then I would be scared that doing this in morph only is not going to be sufficient to avoid the assertion in all cases.

Reply:
It appears in fgMorphArgs via assert prop

```
Morphing args for 8.CALL:

Assertion prop in BB01:
Constant Assertion: V01 == [0000023E8BB78DE0], index = #01
               [000012] H----------                         *  CNS_INT(h) long   0x8BB78DE0 "aaa"

Sorting the arguments:
Deferred argument ('rcx'):
               [000007] n---G+-----                         *  IND       int
               [000012] H----+-----                         \--*  CNS_INT(h) long   0x8BB78DE0 "aaa"
Moved to late list
```

> I would be scared

from my understanding, it's just a debug-only hard requirement of `fgDebugCheckFlags` that GT_IND on top of some specific Icon handles are required to be gtf_invariant

Reply:
https://github.com/dotnet/runtime/blob/8de1809155ea6c32494d5c5816455c7f332ea147/src/coreclr/jit/fgdiagnostic.cpp#L3075-L3078

Reply:
> from my understanding, it's just a debug-only hard requirement of fgDebugCheckFlags that GT_IND on top of some specific Icon handles are required to be gtf_invariant

Yes, I think this is just about missed opportunities, but I imagine there are other situations where we would hit the assert. I wonder how useful this particular flag check is, maybe it should just be removed.

Anyway, we can wait and see if we run into it again.

Reply:
I actually once wanted to remove it but was told I should keep it as long as it doesn't cause issues 😄